### ISSUE TITLE
linear_eq_to_matrix: fails on simple example

### ISSUE THREAD

**Issue by cbm755** [sympy/sympy#16577]
SymPy 1.3 is wrong
```
>>> eqn = Eq(a*(x + 2*y) + b*y, 3)
a⋅(x + 2⋅y) + b⋅y = 3

>>> eqn_expand = eqn.expand()
a⋅x + 2⋅a⋅y + b⋅y = 3

>>> linear_eq_to_matrix([eqn], [x, y])   # WRONG
([0  b], [3])

>>> linear_eq_to_matrix([eqn_expand], [x, y])    # RIGHT
([a  2⋅a + b], [3])
```

SymPy 1.4-rc2 is a different kind of wrong:
```
>>> linear_eq_to_matrix([eqn], [x, y])      # WRONG
([a  b], [3])

>>> linear_eq_to_matrix([eqn_expand], [x, y])   # RIGHT
([a  2⋅a + b], [3])
```

@asmeurer this is sort of a regression, although it was broken in 1.3 as well.  Do you want to hold 1.4 for this or not?

Some effort went into `linear_eq_to_matrix` in this last cycle so b/c of that I'd be inclined to try to fix this; although not sure I'll get to it today...

**Comment by cbm755** [comment]
Simpler example: `eqn = Eq(a*(x + y), 0)`

Traced the problem to here:
https://github.com/sympy/sympy/blob/d15b67d9a62e7528fb18fc29cc5817624a42ad45/sympy/solvers/solveset.py#L1861

@smichr there is a list, we discard everything except the first entry...  Seems wrong...  should we iterate over the list instead?

Also, IMHO, this sort of pattern should be:
```
assert len(list(d1.items()) == 1, "wtf"
list(d1.items())[0]
```
Or otherwise written in a way that an error occurs if the length is unexpected...

**Comment by smichr** [comment]
> IMHO, this sort of pattern should be:

I really would like to see more "bare asserts" like this. They don't need an elaborate error message. They should just stop the program when what the writer thought should be true (but imagined that it could be otherwise) was not true. 

Thanks for the quick fix on this. It looked correct to me and the test that I thought would be good -- one where vf values were non-trivial -- was added.

**Comment by oscarbenjamin** [comment]
> Also, IMHO, this sort of pattern should be:
> 
> ```
> assert len(list(d1.items()) == 1, "wtf"
> list(d1.items())[0]
> ```

I would do that with:
```python
((xf, vf),) = d1.items()
```

**Comment by cbm755** [comment]
Fixed in 1.4 branch by dda9bb62b248583aa725f6316b01cabb36822f8c

**Comment by cbm755** [comment]
@oscarbenjamin yes, re tuple unpacking.  We should watch for "list implicitly assumed to have a particular length" bugs: I think many sympy bugs happen quietly b/c of this when instead they could be raising exceptions.
